### PR TITLE
Paho MQTT library

### PR DESCRIPTION
**Pull request by simonegiacomelli**
These are an incomplete but working header for Paho MQTT javascript library by Eclipse.

I followed this advice
https://discuss.kotlinlang.org/t/kotlin-headers-for-mqtt-javascript-library/6726/5

### CODE REVIEW

File: externals/paho-mqtt/v0/konfig.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
+  "version": "3.1.1"
```
Comment (bashor) [inline externals/paho-mqtt/v0/konfig.json]: Would be nice to rename the directory `v0` according this version, e.g. to `v3`

File: externals/paho-mqtt/v0/src/index.kt [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+            var reconnect: Boolean
+            var useSSL: Boolean
+            var onSuccess: (response: dynamic) -> Unit
+            var onFailure: (response: Response) -> Unit
+        }
+    }
+}
+
+object PahoTest {
+    fun tryit(conf:dynamic) {
```
Comment (bashor) [inline externals/paho-mqtt/v0/src/index.kt]: Minor: missed space before `dynamic`

File: externals/paho-mqtt/v0/src/index.kt [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+            var password: String
+            var reconnect: Boolean
+            var useSSL: Boolean
+            var onSuccess: (response: dynamic) -> Unit
+            var onFailure: (response: Response) -> Unit
+        }
+    }
+}
+
+object PahoTest {
```
Comment (bashor) [inline externals/paho-mqtt/v0/src/index.kt]: API test should be moved to separate directory, see how it's done jquery

File: externals/paho-mqtt/v0/src/index.kt [hunk 4]
```diff
@@ -0,0 +1,63 @@
+//headers for Paho JavaScript Client
+//The Paho JavaScript Client is an MQTT browser-based client library written in Javascript that uses WebSockets to connect to an MQTT Broker.
+//https://github.com/eclipse/paho.mqtt.javascript
+
```
Comment (bashor) [inline externals/paho-mqtt/v0/src/index.kt]: Please add information about yourself, like:
```
// Contributed by: name/nick <link to github user or your email>
```

File: externals/paho-mqtt/v0/src/index.kt [hunk 5]
```diff
@@ -0,0 +1,63 @@
+//headers for Paho JavaScript Client
```
Comment (bashor) [inline externals/paho-mqtt/v0/src/index.kt]: Minor: missed spaces after `//`

File: externals/paho-mqtt/v0/src/index.kt [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+//The Paho JavaScript Client is an MQTT browser-based client library written in Javascript that uses WebSockets to connect to an MQTT Broker.
+//https://github.com/eclipse/paho.mqtt.javascript
+
+package js.externals.paho_mqtt
+
+external object Paho {
```
Comment (bashor) [inline externals/paho-mqtt/v0/src/index.kt]: Have you tried to convert `https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/paho-mqtt/` using [ts2kt](https://github.com/Kotlin/ts2kt/)?

File: externals/paho-mqtt/v0/src/index.kt [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    object MQTT {
+        class Client(hostname: String, port: Int, clientId: String) {
+            val hostname: String
```
Comment (bashor) [inline externals/paho-mqtt/v0/src/index.kt]: According [the docs](http://www.eclipse.org/paho/files/jsdoc/Paho.MQTT.Client.html) it should be called `host`.

File: externals/paho-mqtt/v0/src/index.kt [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+            var onMessageArrived: (msg: Message) -> Unit
+            var onConnectionLost: (response: Response) -> Unit
+        }
+
+        class Message(payloadString: String) {
+            val payloadString: String
+            var destinationName: String
+        }
+
+        class Response {
```
Comment (bashor) [inline externals/paho-mqtt/v0/src/index.kt]: I think it should be an `interface`

File: externals/paho-mqtt/v0/src/index.kt [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+            val errorCode: Int
+            val errorMessage: String
+        }
+
+        interface Options {
```
Comment (bashor) [inline externals/paho-mqtt/v0/src/index.kt]: IMHO `ConnectOptions` is a better name for this, also looks like most of properties can be optional

**Review by bashor — CHANGES_REQUESTED** [review]

**Comment by bashor** [comment]
@simonegiacomelli, thank you for the contribution! (You are first ;))

Would be nice to get some improvements (see comments) before we merge it.

File: externals/paho-mqtt/v0/src/index.kt (hunk 6 above)
Comment (simonegiacomelli) [reply externals/paho-mqtt/v0/src/index.kt]: I tried ts2kt on paho-mqtt typescript but I got an error.
I tried it on jquery typescript and everything was fine... so I think my proceedings were all right.

This is the error I got 
ts2kt version: 0.1.3
Converting index.d.ts
fs.js:640
  return binding.open(pathModule._makeLong(path), stringToFlags(flags), mode);
                 ^

Error: ENOENT: no such file or directory, open 'C:\module.d.ts'
    at Error (native)
    at Object.fs.openSync (fs.js:640:18)
    at Object.fs.readFileSync (fs.js:508:33)
    at getScriptSnapshotFromFile (C:\Users\Simone\AppData\Roaming\npm\node_modules\ts2kt\ts2kt.js:5398:46)
    at translate (C:\Users\Simone\AppData\Roaming\npm\node_modules\ts2kt\ts2kt.js:5623:21)
    at translateToDir (C:\Users\Simone\AppData\Roaming\npm\node_modules\ts2kt\ts2kt.js:5177:26)
    at main (C:\Users\Simone\AppData\Roaming\npm\node_modules\ts2kt\ts2kt.js:5387:5)
    at C:\Users\Simone\AppData\Roaming\npm\node_modules\ts2kt\ts2kt.js:6243:3
    at Object.<anonymous> (C:\Users\Simone\AppData\Roaming\npm\node_modules\ts2kt\ts2kt.js:6246:2)
    at Module._compile (module.js:570:32)

**Comment by simonegiacomelli** [comment]
@bashor I fixed all your requests except the optional parameters of ConnectOptions

File: externals/paho-mqtt/v0/src/index.kt (hunk 6 above)
Comment (bashor) [reply externals/paho-mqtt/v0/src/index.kt]: Could you please [file an issue to Kotlin/ts2kt](https://github.com/Kotlin/ts2kt/issues/new)

**Comment by bashor** [comment]
Nice! Thank you, @simonegiacomelli!

**Pull request merged**